include/Platform/Universal Windows/UWP/ObjCHelperAsyncImplementation.h

```diff
@@ -77,4 +78,57 @@ class AsyncOperationImpl :
     void OnClose() override {}
     void OnCancel() override {}
 };
+
+template<typename resultType, typename resultInterfaceType>
+class AsyncOperationImpl<resultType, resultInterfaceType, true> :
+    public RuntimeClass<
+    RuntimeClassFlags<WinRtClassicComMix>,
+    IAsyncOperation<resultType>,
+    AsyncBase<IAsyncOperationCompletedHandler<resultType>>> {
+
+    typedef RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>,
+        IAsyncOperation<resultType>,
+        AsyncBase<IAsyncOperationCompletedHandler<resultType>>> RuntimeClassT;
+
+    // This will fail for double pointer (and higher level pointers), but we do not care so much about them.
+    InspectableClass((std::wstring(L"Windows.Foundation.IAsyncOperation") + getWString(typeid(std::remove_pointer<resultType>).name())).c_str(), BaseTrust);
+
+    ComPtr<typename std::remove_pointer<resultInterfaceType>::type> _result;
+public:
+    AsyncOperationImpl()
+    {
+        Start();
+    }
+    IFACEMETHODIMP put_Completed(IAsyncOperationCompletedHandler<resultType> *pCompleteHandler) override
+    {
+        return AsyncBase::PutOnComplete(pCompleteHandler);
+    }
+    IFACEMETHODIMP get_Completed(IAsyncOperationCompletedHandler<resultType> **ppCompleteHandler) override
+    {
+        return AsyncBase::GetOnComplete(ppCompleteHandler);
+    }
+    void setResult(resultInterfaceType result)
+    {
+        _result.Attach(result);
+        FireCompletion();
+    }
+    IFACEMETHODIMP GetResults(resultInterfaceType* results) override
+    {
+        *results = 0;
+        HRESULT hr = AsyncBase::CheckValidStateForResultsCall();
+        if (SUCCEEDED(hr))
+        {
+            // Implicit AddRef
+            ComPtr<typename std::remove_pointer<resultInterfaceType>::type> ptr = _result;
+            *results = ptr.Detach();
+        }
+        return hr;
+    }
+    HRESULT OnStart() override
+    {
+        return S_OK;
+    }
+    void OnClose() override {}
+    void OnCancel() override {}
```

Comment:
This is all duplicated.  It would be nice to avoid the duplication.

However, for now, let's leave it.
I'd really like to switch to some WRL templates if they exist.

I'll file a task to:
1. find and swtich to a supported WRL template.
2. If that doesn't exist, clean this up.